File: src/main/java/hantonik/atomic/core/storage/handler/AtomicItemStackHandler.java
package hantonik.atomic.core.storage.handler;

import net.minecraft.core.NonNullList;
import net.minecraft.world.Container;
import net.minecraft.world.SimpleContainer;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.items.ItemStackHandler;
import org.apache.commons.lang3.ArrayUtils;

import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Consumer;

public class AtomicItemStackHandler extends ItemStackHandler {
    private final Consumer<Integer> onContentsChanged;
    private final Map<Integer, Integer> slotSizeMap;
    private BiFunction<Integer, ItemStack, Boolean> slotValidator = null;
    private int maxStackSize = 64;
    private int[] outputSlots = null;

    public AtomicItemStackHandler(int size) {
        this(size, null);
    }

    public AtomicItemStackHandler(int size, Consumer<Integer> onContentsChanged) {
        super(size);

        this.onContentsChanged = onContentsChanged;
        this.slotSizeMap = new HashMap<>();
    }

    @Override
    public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
        if (this.outputSlots != null && ArrayUtils.contains(this.outputSlots, slot))
            return stack;

        return super.insertItem(slot, stack, simulate);
    }

    @Override
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        if (this.outputSlots != null && !ArrayUtils.contains(this.outputSlots, slot))
            return ItemStack.EMPTY;

        return super.extractItem(slot, amount, simulate);
    }

    @Override
    public int getSlotLimit(int slot) {
        return this.slotSizeMap.containsKey(slot) ? this.slotSizeMap.get(slot) : this.maxStackSize;
    }

    @Override
    public boolean isItemValid(int slot, ItemStack stack) {
        return this.slotValidator == null || this.slotValidator.apply(slot, stack);
    }

    @Override
    protected void onContentsChanged(int slot) {
        if (this.onContentsChanged != null)
            this.onContentsChanged.accept(slot);
    }

    public ItemStack insertItemSuper(int slot, ItemStack stack, boolean simulate) {
        return super.insertItem(slot, stack, simulate);
    }

    public ItemStack extractItemSuper(int slot, int amount, boolean simulate) {
        return super.extractItem(slot, amount, simulate);
    }

    public NonNullList<ItemStack> getStacks() {
        return this.stacks;
    }

    public int[] getOutputSlots() {
        return this.outputSlots;
    }

    public void setDefaultSlotLimit(int size) {
        this.maxStackSize = size;
    }

    public void addSlotLimit(int slot, int size) {
        this.slotSizeMap.put(slot, size);
    }

    public void setSlotValidator(BiFunction<Integer, ItemStack, Boolean> validator) {
        this.slotValidator = validator;
    }

    public void setOutputSlots(int... slots) {
        this.outputSlots = slots;
    }

    public Container toIInventory() {
        return new SimpleContainer(this.stacks.toArray(new ItemStack[0]));
    }

    public AtomicItemStackHandler copy() {
        var newInventory = new AtomicItemStackHandler(this.getSlots(), this.onContentsChanged);

        newInventory.setDefaultSlotLimit(this.maxStackSize);
        newInventory.setSlotValidator(this.slotValidator);
        newInventory.setOutputSlots(this.outputSlots);

        this.slotSizeMap.forEach(newInventory::addSlotLimit);

        for (int i = 0; i < this.getSlots(); i++) {
            var stack = this.getStackInSlot(i);

            newInventory.setStackInSlot(i, stack.copy());
        }

        return newInventory;
    }
}

File: Forge/src/main/kotlin/hantonik/atomic/core/storage/handler/AtomicItemStackHandler.kt
package hantonik.atomic.core.storage.handler

import net.minecraft.core.NonNullList
import net.minecraft.world.Container
import net.minecraft.world.SimpleContainer
import net.minecraft.world.item.ItemStack
import net.minecraftforge.items.ItemStackHandler

open class AtomicItemStackHandler(size: Int, private val onContentsChanged: ((Int) -> Unit)?) : ItemStackHandler(size) {
    private val slotSizeMap: MutableMap<Int, Int> = mutableMapOf()
    private var slotValidator: ((Int, ItemStack) -> Boolean)? = null
    private var maxStackSize = 64

    var outputsSlots: IntArray? = null

    constructor(size: Int) : this(size, null)

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        if (this.outputsSlots != null && this.outputsSlots!!.contains(slot))
            return stack

        return super.insertItem(slot, stack, simulate)
    }

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        if (this.outputsSlots != null && this.outputsSlots!!.contains(slot))
            return ItemStack.EMPTY

        return super.extractItem(slot, amount, simulate)
    }

    override fun getSlotLimit(slot: Int): Int {
        return if (this.slotSizeMap.containsKey(slot))
            this.slotSizeMap[slot]!!
        else
            this.maxStackSize
    }

    override fun isItemValid(slot: Int, stack: ItemStack) = this.slotValidator == null || this.slotValidator!!.invoke(slot, stack)

    override fun onContentsChanged(slot: Int) {
        if (this.onContentsChanged != null)
            this.onContentsChanged.invoke(slot)
    }

    fun insertItemSuper(slot: Int, stack: ItemStack, simulate: Boolean) = super.insertItem(slot, stack, simulate)

    fun extractItemSuper(slot: Int, amount: Int, simulate: Boolean) = super.extractItem(slot, amount, simulate)

    fun getStacks(): NonNullList<ItemStack> = this.stacks

    fun setDefaultSlotLimit(size: Int) {
        this.maxStackSize = size
    }

    fun addSlotLimit(slot: Int, size: Int) {
        this.slotSizeMap[slot] = size
    }

    fun setSlotValidator(validator: ((Int, ItemStack) -> Boolean)?) {
        this.slotValidator = validator
    }

    fun setOutputSlots(vararg slots: Int) {
        this.outputsSlots = slots
    }

    fun toIInventory(): Container {
        return SimpleContainer(*this.stacks.toTypedArray())
    }

    fun copy(): AtomicItemStackHandler {
        val copy = AtomicItemStackHandler(this.slots, this.onContentsChanged)

        copy.setDefaultSlotLimit(this.maxStackSize)
        copy.setSlotValidator(this.slotValidator)

        if (this.outputsSlots != null)
            copy.setOutputSlots(*this.outputsSlots!!)

        this.slotSizeMap.forEach(copy::addSlotLimit)

        for (index in 0 until this.slots)
            copy.setStackInSlot(index, this.getStackInSlot(index))

        return copy
    }
}